lib/services/embeddings.go

```diff
@@ -0,0 +1,396 @@
+	// the stale data
+	if n.stale.Load() {
+		_ = n.getResourcesAndUpdateCurrent(ctx)
+	}
+
+	needsEmbedding := make(map[string][]byte)
+	n.mutex.Lock()
+	timestamp := time.Now()
+	for nodeName, node := range n.currentNodes {
+		if node.needsEmbedding {
+			text, err := serializeNode(node.node)
+			if err != nil {
+				n.Log.Warningf("failed to serialize node %s, the node won't be embedded", node.node.GetName())
+				continue
+			}
+			needsEmbedding[nodeName] = text
+		}
+	}
+	n.mutex.Unlock()
+
+	embeddings, err := n.Embed(ctx, types.KindNode, needsEmbedding)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	for _, embedding := range embeddings {
+		if node, ok := n.currentNodes[embedding.GetEmbeddedID()]; ok && node.lastUpdate.Before(timestamp) {
+			node.gotEmbedded()
+		}
+	}
+
+	n.Log.Debugf("Embedded %d nodes", len(embeddings))
+
+	// TODO(hugoShaka): when vector index is here, delete then insert nodes in it.
+	return nil
+}
+
+func (n *nodeEmbeddingCollector) getNodes(ctx context.Context) (map[string]types.Server, error) {
+	nodes, err := n.NodesGetter.GetNodes(ctx, apidefaults.Namespace)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(nodes) == 0 {
+		return map[string]types.Server{}, nil
+	}
+
+	current := make(map[string]types.Server, len(nodes))
+	for _, node := range nodes {
+		current[node.GetName()] = node
+	}
+
+	return current, nil
+}
+
+func (n *nodeEmbeddingCollector) defineCollectorAsInitialized() {
+	n.once.Do(func() {
+		// mark watcher as initialized.
+		close(n.initializationC)
+	})
+}
+
+// processEventAndUpdateCurrent is called when a watcher event is received.
+func (n *nodeEmbeddingCollector) processEventAndUpdateCurrent(_ context.Context, event types.Event) {
+	if event.Resource == nil || event.Resource.GetKind() != types.KindNode {
+		n.Log.Warningf("Unexpected event: %v.", event)
+		return
+	}
+
+	n.mutex.Lock()
+	timestamp := time.Now()
+	defer n.mutex.Unlock()
+	switch event.Type {
+	case types.OpDelete:
+		n.removeNodes([]string{event.Resource.GetName()}, timestamp)
+	case types.OpPut:
+		server, ok := event.Resource.(types.Server)
+		if !ok {
+			n.Log.Warningf("Unexpected type %T.", event.Resource)
+			return
+		}
+		n.addNodes(map[string]types.Server{server.GetName(): server}, timestamp)
+	default:
+		n.Log.Warningf("Skipping unsupported event type %s.", event.Type)
+	}
+}
+
+func (n *nodeEmbeddingCollector) initializationChan() <-chan struct{} {
+	return n.initializationC
+}
+
+func (n *nodeEmbeddingCollector) notifyStale() {
+	n.stale.Store(true)
+}
+
+// NodeCount returns how many nodes are tracked by the collector and if they
+// need embedding.
+func (n *nodeEmbeddingCollector) NodeCount(needsEmbedding bool) int {
+	count := 0
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	for _, node := range n.currentNodes {
+		if node.needsEmbedding == needsEmbedding {
+			count += 1
+		}
+	}
+	return count
+}
```

Comment:
We can optimize this function, but I think it would defeat this counter's purpose. Tests need a way to peak into the collector's state to know what is happening (and the tests are living in a separate `services_test` package, a private function is not usable here). Making this counter smarter and more efficient increases the risks of it returning an inaccurate value because of some collector bug.